Library/Homebrew/extend/os/mac/keg_relocate.rb

```diff
@@ -137,6 +137,15 @@ def find_dylib(bad_name)
     lib.find { |pn| break pn if pn.to_s.end_with?(suffix) }
   end
 
+  def relocate_build_prefix(old_cellar, new_prefix, new_cellar)
+    macho_files.each do |file|
```

Comment:
I can't remember if we already do this but if we don't: we should store something in the `Tab`/`INSTALL_RECEIPT.json` of which files need to be relocated so we can avoid every `pour` for every user having to `scan` every ELF/MachO file when we can know at bottling time which ones will need/not need relocated.

Reply:
I had the exact same thought and this relates the above question about whether we scan all ELF/MachO files at some other time.  We do scan all of them when we bottle, but it appears the files where build prefix references were found aren't actually saved anywhere, just printed as output of `brew bottle`.  This would obviously require rebottling, but we will already need to do this to get a longer install prefix so it would probably make sense to handle both issues at the same time.

Reply:
This doesn't need to block this PR in this case but I think it's a blocker before we would roll this out by default. Rebottling to improve performance for end-users here seems like big win.

Reply:
That sounds like a good plan.  We will have to rebottle all the non-relocatable bottles anyway and while we are at it, we can store the files that will need to be relocated in the bottle spec.  This is exactly how Anaconda handles this, probably for the same reason.    This would also allow us to record non-ELF/MachO binaries which may need relocation, some of which don't require null padding i.e. GLib GIR files.  

In the interim, I got some performance gain by using `keg.each_unique_file_matching` to scan all the files first, and then only read in binaries which were returned by the list.  I think this is faster because `keg.each_unique_file_matching` uses `fgrep` to find the strings instead of reading the binaries in with `File.binread`.  However this should be superseded by just using the list of files which were identified as needing relocation during bottling.

Reply:
> In the interim, I got some performance gain by using keg.each_unique_file_matching to scan all the files first, and then only read in binaries which were returned by the list. I think this is faster because keg.each_unique_file_matching uses fgrep to find the strings instead of reading the binaries in with File.binread. However this should be superseded by just using the list of files which were identified as needing relocation during bottling.

Yeh, seems like a good interim speedup. I'd suggest memoizing the results and/or ensuring it isn't run twice.

Reply:
As it's written now, the results are stored in a list and that is what is now iterated for the actual patching.  So the binaries are only checked for the strings once before patching, and I will be adding to check them again after patching, in both cases using `keg.each_unique_file_matching`.

Reply:
> in both cases using keg.each_unique_file_matching.

This doesn't seem to be memoized so if we're calling it more than once (or could or expect we might in future): it'd be good to do that.

Reply:
>This doesn't seem to be memoized so if we're calling it more than once (or could or expect we might in future): it'd be good to do that.

The only thing I'm unsure about here is that `keg.each_unique_file_matching` just takes a string and a block (because it has a `yield` statement) and returns the files that match the string.  It doesn't seem like we have a method that takes an array of file paths and scans them for a string with `fgrep`, which I think is what we'd want for the second check.  

The goal of scanning the binaries for the prefix again after patching is mostly to catch binaries which are not ELF/MachO, because those are the only binaries we patch now.  As a sanity check we also want to double check the ELF/MachO binaries as well.  Ideally we should only scan those files and ignore all the others, because they didn't have a match the first time.  

If `keg.each_unique_file_matching` is not suitable for that second check, maybe we need to write a new method that does take a list of files and scans them individually with `fgrep` instead?

Reply:
> Ideally we should only scan those files and ignore all the others, because they didn't have a match the first time.

Yeh, agreed. Basically I just want to avoid running `fgrep` multiple times on the same file if we can avoid it, however that is done.

Note: that does not need to block this PR.

Reply:
Okay, I'll mark this as resolved for now.  Regardless of whether we change this now, once we start recording files that need relocation during bottling, we'll need a method that scans a specific list of files with `fgrep` because we'll already know which files need relocation.  I think it would make sense to tackle that when we've made those changes to the bottles.